server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -0,0 +1,465 @@
+    private final LocalCheckpointTracker localCheckpointTracker;
+    private final Translog translog;
+
+    public NRTReplicationEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+        store.incRef();
+        NRTReplicationReaderManager readerManager = null;
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            readerManager = new NRTReplicationReaderManager(OpenSearchDirectoryReader.wrap(getDirectoryReader(), shardId));
+            final SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+                this.lastCommittedSegmentInfos.getUserData().entrySet()
+            );
+            this.localCheckpointTracker = new LocalCheckpointTracker(commitInfo.maxSeqNo, commitInfo.localCheckpoint);
+            this.completionStatsCache = new CompletionStatsCache(() -> acquireSearcher("completion_stats"));
+            this.readerManager = readerManager;
+            this.readerManager.addListener(completionStatsCache);
+            this.translog = openTranslog(
+                engineConfig,
+                getTranslogDeletionPolicy(engineConfig),
+                engineConfig.getGlobalCheckpointSupplier(),
+                localCheckpointTracker::markSeqNoAsPersisted
+            );
+        } catch (IOException e) {
+            IOUtils.closeWhileHandlingException(store::decRef, readerManager);
+            throw new EngineCreationFailureException(shardId, "failed to create engine", e);
+        }
+    }
+
+    public synchronized void updateSegments(final SegmentInfos infos, long seqNo) throws IOException {
+        // Update the current infos reference on the Engine's reader.
+        readerManager.updateSegments(infos);
+
+        // only update the persistedSeqNo and "lastCommitted" infos reference if the incoming segments have a higher
+        // generation. We can still refresh with incoming SegmentInfos that are not part of a commit point.
+        if (infos.getGeneration() > lastCommittedSegmentInfos.getGeneration()) {
+            this.lastCommittedSegmentInfos = infos;
+        }
+        localCheckpointTracker.fastForwardProcessedSeqNo(seqNo);
+        readerManager.maybeRefresh();
+    }
+
+    @Override
+    public String getHistoryUUID() {
+        return loadHistoryUUID(lastCommittedSegmentInfos.userData);
+    }
+
+    @Override
+    public long getWritingBytes() {
+        return 0;
+    }
+
+    @Override
+    public CompletionStats completionStats(String... fieldNamePatterns) {
+        return completionStatsCache.get(fieldNamePatterns);
+    }
+
+    @Override
+    public long getIndexThrottleTimeInMillis() {
+        return 0;
+    }
+
+    @Override
+    public boolean isThrottled() {
+        return false;
+    }
+
+    @Override
+    public void trimOperationsFromTranslog(long belowTerm, long aboveSeqNo) throws EngineException {
+        ensureOpen();
```

Comment:
Why we check for `ensureOpen()` here and not in any other place where `translog` is involved? (add/delete/...)

Reply:
An oversight, added to index/delete/noOp.